src/GraphNavigator/DeserializationGraphNavigator.php

```diff
@@ -107,9 +108,24 @@ public function accept($data, ?array $type = null)
         if (null === $type) {
             throw new RuntimeException('The type must be given for all properties when deserializing.');
         }
+
+        $shouldDeserializeNull = false;
+        $metadataStack = $this->context->getMetadataStack();
+
+        if (!$metadataStack->isEmpty()) {
+            $currentPropertyMetadata = $metadataStack->top();
+            if ($currentPropertyMetadata instanceof PropertyMetadata) {
+                $shouldDeserializeNull = $currentPropertyMetadata->deserializeNull;
+            }
+        }
+
         // Sometimes data can convey null but is not of a null type.
         // Visitors can have the power to add this custom null evaluation
-        if ($this->visitor instanceof NullAwareVisitorInterface && true === $this->visitor->isNull($data)) {
+        // If null is explicitly allowed we should skip this
+        if ($this->visitor instanceof NullAwareVisitorInterface
+            && true === $this->visitor->isNull($data)
+            && !$shouldDeserializeNull
```

Comment:
Can you move `!$shouldDeserializeNull` as first condition? so in case is true we do not need to call `isNull`